remote-control/luna-control.man
.\" This file is part of the Lightweight Universal Network Analyzer (LUNA)
.\"
.\" Copyright (c) 2014 Fiona Klute
.\"
.\" LUNA is free software: you can redistribute it and/or modify it
.\" under the terms of the GNU General Public License as published by
.\" the Free Software Foundation, either version 3 of the License, or
.\" (at your option) any later version.
.\"
.\" LUNA is distributed in the hope that it will be useful, but WITHOUT
.\" ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
.\" or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
.\" License for more details.
.\"
.\" You should have received a copy of the GNU General Public License
.\" along with LUNA. If not, see <http://www.gnu.org/licenses/>.
.TH LUNA-CONTROL 1 2014-06-16 "LUNA" "LUNA Manual"

.SH NAME
luna-control \- run one or more LUNA transmissions as defined in a
configuration file

.SH SYNOPSIS
.B luna-control
CONFFILE

.SH DESCRIPTION
.P
LUNA, short for "Lightweight Universal Network Analyzer", is a system
to analyze network behavior under different network loads. The
.B luna-control
command provides a way to automate LUNA measurements using a simple
configuration file (see section
.B CONFIGURATION FILE
below). It requires the low-level
.BR luna (1)
command to be available on the hosts configured as clients and
servers. The host running \fBluna-control\fR is called the \fBcontrol
host\fR.

.SH OPTIONS

None.

.SH CONFIGURATION FILE
.P
The configuration file consists of variable assignments in
.RS
.P
NAME=VALUE
.RE
.P
form. Additionally, the file is split in sections. The top of the
configuration file before any explicit section definition is the
global section, which contains common settings and defaults. All
further sections are connection sections, each of which configures one
LUNA transmission. A section definition has the following form:
.RS
.P
[NAME]
.RE
.P
All variables after the section definition belong to that section, up
to either another section definition or the end of file.

Names of variables and sections may contain letters, numbers and
underscores (\fB\\w\fR in Perl regexp syntax). Empty lines and lines
starting with \fB#\fR are ignored. Section headers must be the only
thing in their line. Variable assignments may be indented if desired,
but whitespace at line ends is prohibited.

Valid variables for global and connection sections are listed
below. Required variables are marked as such, others are optional and
will use default values if not set. An example configuration file is
shown in the section
.B EXAMPLE
below.

.SS Global Settings

These variables must be set before any connection sections.

.TP
.B default_exec
The default path to the \fBluna\fR binary on client and server hosts,
can be overridden for individual hosts using \fBclient_exec\fR and
\fBserver_exec\fR. Default value: /usr/local/bin/luna

.TP
.B startup_delay
If \fBstartup_delay\fR is set, \fBluna-control\fR will configure
clients to start sending at a fixed time. This time is calculated as
the time on the control host right before starting connection
handlers, plus \fBstartup_delay\fR seconds, modified by the
connection-specific \fBdelay\fR (if any). Note that LUNA relies on
each host's clock for time measurement, so the quality of the
synchronization depends on proper clock synchronization across
hosts. The value of \fBstartup_delay\fR must be large enough to
accommodate SSH connection setup to both client and server hosts to be
effective. If not set, each connection will be started as soon as
possible.

.SS Connection Section

Each connection section may contain the following variables:

.TP
.B server
The LUNA server host (traffic sink, may send echo packets if
requested). Required.

.TP
.B client
The LUNA client host (traffic source). Required.

.TP
.B port
The listening port on the server, default is 4567. If your
configuration contains multiple connections with the server running on
the same host, you must configure all but one of the servers to use
another port, or connections will fail because the server cannot bind
its listening port.

.TP
.B target
By default, the LUNA client will use the IP address or hostname
defined in \fBserver\fR as the transmission target. If it should use a
different IP address or hostname (e.g. because the experiment uses a
separate control network), set the server to pass to the LUNA client
here.

.TP
.B generator
The generator the client will use, if not set, the default generator
will be used. All generators supported by the LUNA version installed
on the client can be used. If an invalid generator name is set, this
connection will fail.

.TP
.B generator_args
Arguments for the configured generator. Valid arguments depend on the
generator.

.TP
.B time
The duration for which the client should send, in seconds. Required.

.TP
.B echo
Set to true if the client should request echo packets.

.TP
.B server_exec
The path to the \fBluna\fR binary on the server, if not set,
\fBdefault_exec\fR is used.

.TP
.B client_exec
The path to the \fBluna\fR binary on the client, if not set,
\fBdefault_exec\fR is used.

.TP
.B server_output
File name on the control host for the server output. Required.

.TP
.B client_output
File name on the control host for the client output. Required if
\fBecho\fR is set to true.

.TP
.B delay
Connection specific delay until transmission start, relative to the
common start time calculated from \fBstartup_delay\fR. If
\fBstartup_delay\fR is not set, \fBdelay\fR has no effect.

.SH EXIT STATUS
.P
.B 0
if execution completed normally, non-zero in case of an error. In the
latter case,
.B luna-control
will output an error message for the user.

.SH ENVIRONMENT
.P
\fBluna-control\fR uses SSH to connect to the hosts defined as clients
and servers. It is highly recommended to configure public key login,
otherwise you will have to enter passwords for each SSH connection.

The Perl module \fBNet::OpenSSH\fR is required on the control host to
manage SSH connections. The Debian style
.BR start-stop-daemon (8)
command is required on all hosts configured as servers.

.SH NOTES

.P
For notes on how to install the low level LUNA tool on the client and
server hosts, please see the
.BR luna (1)
manual page and the README. All further configuration for using
\fBluna-control\fR is done on the control host only.

.SH EXAMPLE
.P
A valid configuration file may look like this:

.RS
.nf
.sp
default_exec=/usr/local/bin/luna
[conn]
client=192.0.2.7
server=192.0.2.8
server_output=test-server.log
client_output=test-client.log
echo=true
generator=gaussian
generator_args=max=400,sigma=30
.fi
.RE

The resulting command on the client (minus output settings) would be:
.RS
.P
luna -c 192.0.2.8 -e -g gaussian -a max=400,sigma=30
.RE
.P
During the measurement, output is written to temporary files on the
hosts and transferred to the control host after the transmission is
complete. Client output is only written if echo is enabled.

.SH SEE ALSO
.P
.BR luna (1),
.BR ssh (1)


remote-control/luna-control.pl
#!/usr/bin/perl

# This file is part of the Lightweight Universal Network Analyzer (LUNA)
#
# Copyright (c) 2013 Fiona Klute
#
# LUNA is free software: you can redistribute it and/or modify it
# under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# LUNA is distributed in the hope that it will be useful, but WITHOUT
# ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
# or FITNESS FOR A PARTICULAR PURPOSE. See the GNU General Public
# License for more details.
#
# You should have received a copy of the GNU General Public License
# along with LUNA. If not, see <http://www.gnu.org/licenses/>.

use strict;
use threads;
use threads::shared;
use Net::OpenSSH;

my @conns;
my @workers;
my %general = (
    "default_exec" => "/usr/local/bin/luna",
    "startup_delay" => undef
    );

my $conffile = $ARGV[0];
die "Configuration file missing or not readable!\n" if (! -r $conffile);

my $rw = open(CONF, $conffile);
die "Could not open configuration file: $!\n" if(not defined($rw));

# parse configuration
my $current = undef;
my $c = 0;
while(defined(my $i = <CONF>))
{
    $c++;
    # ignore comments and empty lines
    next if ($i =~ /^\#.*$/ or $i =~ /^$/);

    # check for section header
    if ($i =~ /\[(\w+)\]/)
    {
	$current = &init_conn($1);
	push(@conns, $current);
	next;
    }

    # normal variable assignment
    if ($i =~ /^\s*(\w+)=(.*)$/)
    {
	if (not defined $current)
	{
	    $general{$1} = $2;
	}
	else
	{
	    $current->{$1} = $2;
	}
	next;
    }

    # Anything unknown, should never happen with a valid configuration file
    chomp($i);
    die "Configuration file line ".$c." is invalid: \"".$i."\"\n";
}
close(CONF);

# show configuration (just for development)
foreach my $i (keys(%general))
{
    print $i."=".$general{$i}."\n";
}
foreach my $conn (@conns)
{
    print "\n";
    foreach my $i (keys(%{$conn}))
    {
	print $i."=".$conn->{$i}."\n";
    }
}

# Set up start time: Start time is "now" plus the configured startup
# delay. Connections may have their own additional delay.
if (defined $general{startup_delay})
{
    my $start_time = time();
    $start_time = $start_time + $general{startup_delay};
    print "General start time: ".$start_time."\n";
    foreach my $conn (@conns)
    {
	$conn->{start_time} = $start_time + $conn->{delay};
    }
}

# Start a worker thread for each connection.
for (my $i = 0; $i < @conns; $i++)
{
     $workers[$i] = threads->create({'context' => 'scalar'},
				    \&run_connection,
				    $conns[$i]);
}

# Wait until all worker threads have terminated, order doesn't matter.
my $ret = 0;
foreach my $thr (@workers)
{
    $ret = $ret + $thr->join();
}
exit($ret);



# This function runs one connection, passed as a hash reference.
sub run_connection
{
    my $conn = $_[0];
    if (not defined $conn->{target})
    {
	$conn->{target} = $conn->{server};
    }

    # start SSH connection to the server host
    $conn->{server_ssh} = &start_ssh($conn->{server});
    # if server and client are on the same host, just reuse the SSH
    # connection, otherwise start SSH connection to the client host
    if ($conn->{server} eq $conn->{client})
    {
	$conn->{client_ssh} = $conn->{server_ssh};
    }
    else
    {
	$conn->{client_ssh} = &start_ssh($conn->{client});
    }
    # Running these three functions synchronously ensures that the
    # server is (most likely) ready when the client starts running,
    # and the transmission is (certainly) complete before the server
    # is stopped.
    my $ret = 0;
    $ret += &start_server($conn);
    $ret += &run_client($conn);
    $ret += &stop_server($conn);
    return $ret;
}



# Establish an SSH connection to the given host, passed as a string
# parameter in the usual format: [user@]server[:sshport]
sub start_ssh
{
    my $host = $_[0];
    my $ssh = Net::OpenSSH->new($host);
    $ssh->error and
	die "Couldn't establish SSH connection with ".$host.": ". $ssh->error;
    return $ssh;
}



# Start the server for this connection
# One parameter: a connection hash reference
sub start_server
{
    my $conn = $_[0];
    my $ssh = $conn->{server_ssh};
    my $ret = 0;

    # create PID file on the server
    $conn->{server_pidfile} = $ssh->capture("mktemp");
    if ($ssh->error)
    {
	$ret = 1;
	print "Could not create server PID file.\n";
    }
    chomp($conn->{server_pidfile});
    # create output file on the server
    $conn->{server_outfile} = $ssh->capture("mktemp");
    if ($ssh->error)
    {
	$ret = 1;
	print "Could not create server output file.\n";
    }
    chomp($conn->{server_outfile});

    # build server start command
    my $command = "/sbin/start-stop-daemon --start";
    # set PID file
    $command = $command." --pidfile ".$conn->{server_pidfile};
    $command = $command." --exec ".$conn->{server_exec}." --background --make-pidfile --";
    # server parameters
    $command = $command." -s -T -p ".$conn->{port}." -o ".$conn->{server_outfile};
    $ssh->system($command);
    if ($ssh->error)
    {
	$ret = 1;
	print "Command \"".$command."\" failed: ". $ssh->error."\n";
    }

    return $ret;
}



# Stop the server for this connection
# One parameter: a connection hash reference
sub stop_server
{
    my $conn = $_[0];
    my $ssh = $conn->{server_ssh};
    my $ret = 0;

    # stop the server, using the stored PID
    my $command = "kill -TERM \$(cat ".$conn->{server_pidfile}.")";
    $ssh->system($command);
    if ($ssh->error)
    {
	$ret = 1;
	print "Command \"".$command."\" failed: ". $ssh->error."\n";
    }

    # delete PID file
    my $command = "rm ".$conn->{server_pidfile};
    $ssh->system($command);
    if ($ssh->error)
    {
	$ret = 1;
	print "Command \"".$command."\" failed: ". $ssh->error."\n";
    }

    # copy output to local storage
    $ssh->scp_get({async => 0}, $conn->{server_outfile},
		  $conn->{server_output});
    if ($ssh->error)
    {
	$ret = 1;
	print "Copying server output to ".$conn->{server_output}." failed: ".$ssh->error."\n";
    }

    # delete output file on the server
    my $command = "rm ".$conn->{server_outfile};
    $ssh->system($command);
    if ($ssh->error)
    {
	$ret = 1;
	print "Command \"".$command."\" failed: ". $ssh->error."\n";
    }

    return $ret;
}



# Run the client process and transfer the output, if echo was
# requested.
# One parameter: a connection hash reference
sub run_client
{
    my $conn = $_[0];
    my $ssh = $conn->{client_ssh};

    # build client command
    my $command = $conn->{client_exec}." -c ".$conn->{target}.
	" -p ".$conn->{port}." -t ".$conn->{time};

    # add -e and output file, if requested
    if ($conn->{echo} eq "true")
    {
	# crate output file
	$conn->{client_outfile} = $ssh->capture("mktemp");
	die "Could not create client output file: " if $ssh->error;
	chomp($conn->{client_outfile});
	$command = $command." -e -o ".$conn->{client_outfile};
    }

    # add generator settings, if present
    if (defined $conn->{generator})
    {
	$command = $command." -g ".$conn->{generator};
	if (defined $conn->{generator_args})
	{
	    $command = $command." -a ".$conn->{generator_args};
	}
    }

    # add start time, if defined
    if (defined $conn->{start_time})
    {
	$command = $command." --start-time=".$conn->{start_time};
    }

    # run client
    $ssh->system($command);
    print "Command \"".$command."\" failed: ". $ssh->error."\n"
	if ($ssh->error);

    # get output, if any
    if (defined $conn->{client_outfile})
    {
	$ssh->scp_get({async => 0}, $conn->{client_outfile}, $conn->{client_output});
	print "Copying client output to ".$conn->{client_output}." failed: ".$ssh->error."\n" if ($ssh->error);

	my $command = "rm ".$conn->{client_outfile};
	$ssh->system($command);
	if ($ssh->error)
	{
	    print "Command \"".$command."\" failed: ". $ssh->error."\n";
	    return 1;
	}
    }
}



# Intialize connection with default values, as far as possible. The
# first and only parameter is the connection identifier (section name
# in the configuration file).
sub init_conn
{
    my %conn =
	(
	 "id" => $_[0],
	 # The following values are configuration options:
	 "server" => undef,
	 "client" => undef,
	 "port" => 4567,
	 # If the client should use a different IP/hostname as the
	 # transmission target than the one specified in server, set
	 # it here.
	 "target" => undef,
	 # generator settings for the client
	 "generator" => undef,
	 "generator_args" => undef,
	 # time for the client to run, in seconds
	 "time" => 2,
	 # set to true to request echo packets
	 "echo" => undef,
	 # commands to run
	 "server_exec" => $general{default_exec},
	 "client_exec" => $general{default_exec},
	 # local output files
	 "server_output" => undef,
	 "client_output" => undef,
	 # connection specific delay, additional to $general{startup_delay}
	 "delay" => 0,
	 # Variables below this point are internal, not for configuration
	 "server_ssh" => undef,
	 "server_pidfile" => undef,
	 "server_outfile" => undef,
	 "client_ssh" => undef,
	 "client_outfile" => undef,
	 "start_time" => undef,
	);
    return \%conn;
}
